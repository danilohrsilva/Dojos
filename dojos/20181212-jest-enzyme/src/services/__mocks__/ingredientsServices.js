export const getIngredients = jest.fn(() => Promise.reject({}));
