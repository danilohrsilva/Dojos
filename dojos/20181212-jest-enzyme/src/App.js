import React, { Component, Fragment } from 'react';

import CartSummaryModal from './Components/CartSummaryModal/CartSummaryModal';
import Ingredient from './Components/Ingredient/Ingredient';
import { getIngredients } from './services/ingredientsServices';

import './App.css';

class App extends Component {
  constructor() {
    super();

    this.state = {
      cart: {
        items: [],
        total: 0,
      },
      isOrderSummaryOpen: false,
      ingredients: [],
    }
  }

  static getDerivedStateFromProps(props, state) {
    const { cart } = state;

    if (!cart.items.length) {
      return {
        isOrderSummaryOpen: false,
      }
    }

    return null;
  }

  componentDidMount() {
    getIngredients()
      .then(ingredients => {
        console.log(ingredients);
        this.setState({
          ingredients,
        });
      })
  }

  addIngredientToCart = (ingredient) => {
    const { cart } = this.state;
    const items = [
      ...cart.items,
      ingredient,
    ];

    const total = items.reduce((total, item) => item.price + total, 0);

    this.setState({
      cart: {
        items,
        total,
      }
    })
  }

  removeIngredientFromCart = (ingredient) => {
    const { cart } = this.state;
    const items = cart.items.filter(item => item.id !== ingredient.id);
    const total = items.reduce((total, item) => item.price + total, 0);

    this.setState({
      cart: {
        items,
        total,
      }
    })
  }

  openOrderSummary = () => {
    this.setState({
      isOrderSummaryOpen: true,
    })
  }

  closeOrderSummary = () => {
    this.setState({
      isOrderSummaryOpen: false,
    })
  }

  render() {
    const isCartEmpty = !this.state.cart.items.length;
    return (
      <Fragment>
        <button
          type="button"
          onClick={this.openOrderSummary}
          disabled={isCartEmpty}
          className="summary-button"
        >
          Resumo do pedido
        </button>

        {this.state.ingredients.map(ingredient => (
          <Ingredient
            cart={this.state.cart}
            ingredient={ingredient}
            addItem={this.addIngredientToCart}
            key={ingredient.id}
          />
        ))}

        <p>Total de itens: {this.state.ingredients.length}</p>

        <CartSummaryModal
          isOpen={this.state.isOrderSummaryOpen}
          onClose={this.closeOrderSummary}
          cart={this.state.cart}
          removeItem={this.removeIngredientFromCart}
        />
      </Fragment>
    );
  }
}

export default App;
