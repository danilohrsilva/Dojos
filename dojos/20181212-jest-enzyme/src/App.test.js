import React from 'react';
import ReactDOM from 'react-dom';
import { shallow, mount } from 'enzyme';
import { getIngredients } from './services/ingredientsServices';

import App from './App';

jest.mock('./services/ingredientsServices');


it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<App />, div);
  ReactDOM.unmountComponentAtNode(div);
});

it('renders without crashing', () => {
  shallow(<App />);
});

it('Teste snapshot', () => {
  const wrapper = shallow(<App />);
  expect(wrapper).toMatchSnapshot();
});

it('fetch on componentDidMount', () => {
  shallow(<App />);
  expect(getIngredients).toHaveBeenCalled();
});

it('error in render', async () => {
  const wrapper = await mount(<App/>);
  expect(wrapper.state().ingredients.length).toBe(0);
})

