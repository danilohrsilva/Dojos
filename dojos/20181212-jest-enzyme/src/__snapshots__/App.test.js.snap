// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Teste snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <button
          className="summary-button"
          disabled={true}
          onClick={[Function]}
          type="button"
        >
          Resumo do pedido
        </button>,
        Array [],
        <p>
          Total de itens: 
          0
        </p>,
        <Component
          cart={
            Object {
              "items": Array [],
              "total": 0,
            }
          }
          isOpen={false}
          onClose={[Function]}
          removeItem={[Function]}
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Resumo do pedido",
          "className": "summary-button",
          "disabled": true,
          "onClick": [Function],
          "type": "button",
        },
        "ref": null,
        "rendered": "Resumo do pedido",
        "type": "button",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Total de itens: ",
            0,
          ],
        },
        "ref": null,
        "rendered": Array [
          "Total de itens: ",
          0,
        ],
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "cart": Object {
            "items": Array [],
            "total": 0,
          },
          "isOpen": false,
          "onClose": [Function],
          "removeItem": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <button
            className="summary-button"
            disabled={true}
            onClick={[Function]}
            type="button"
          >
            Resumo do pedido
          </button>,
          Array [],
          <p>
            Total de itens: 
            0
          </p>,
          <Component
            cart={
              Object {
                "items": Array [],
                "total": 0,
              }
            }
            isOpen={false}
            onClose={[Function]}
            removeItem={[Function]}
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Resumo do pedido",
            "className": "summary-button",
            "disabled": true,
            "onClick": [Function],
            "type": "button",
          },
          "ref": null,
          "rendered": "Resumo do pedido",
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Total de itens: ",
              0,
            ],
          },
          "ref": null,
          "rendered": Array [
            "Total de itens: ",
            0,
          ],
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "cart": Object {
              "items": Array [],
              "total": 0,
            },
            "isOpen": false,
            "onClose": [Function],
            "removeItem": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
